cursive-core/src/views/checkbox.rs

```diff
@@ -1,14 +1,122 @@
+use ahash::{HashSet, HashSetExt};
+
 use crate::{
     direction::Direction,
     event::{Event, EventResult, Key, MouseButton, MouseEvent},
     style::PaletteStyle,
+    utils::markup::StyledString,
     view::{CannotFocus, View},
     Cursive, Printer, Vec2, With,
 };
+use parking_lot::Mutex;
+use std::hash::Hash;
 use std::sync::Arc;
 
+type GroupCallback<T> = dyn Fn(&mut Cursive, &HashSet<Arc<T>>) + Send + Sync;
 type Callback = dyn Fn(&mut Cursive, bool) + Send + Sync;
 
+struct SharedState<T> {
+    selections: HashSet<Arc<T>>,
+    values: Vec<Arc<T>>,
+
+    on_change: Option<Arc<GroupCallback<T>>>,
+}
+
+impl<T> SharedState<T> {
+    pub fn selections(&self) -> &HashSet<Arc<T>> {
+        &self.selections
+    }
+}
+
+/// Group to coordinate multiple checkboxes.
+///
+/// A `MultiChoiceGroup` can be used to create and manage multiple [`Checkbox`]es.
+///
+/// A `MultiChoiceGroup` can be cloned; it will keep shared state (pointing to the same group).
+pub struct MultiChoiceGroup<T> {
+    // Given to every child button
+    state: Arc<Mutex<SharedState<T>>>,
+}
+
+// We have to manually implement Clone.
+// Using derive(Clone) would add am unwanted `T: Clone` where-clause.
+impl<T> Clone for MultiChoiceGroup<T> {
+    fn clone(&self) -> Self {
+        Self {
+            state: Arc::clone(&self.state),
+        }
+    }
+}
+
+impl<T: 'static + Hash + Eq> Default for MultiChoiceGroup<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: 'static + Hash + Eq> MultiChoiceGroup<T> {
+    /// Creates an empty group for check boxes.
+    pub fn new() -> Self {
+        Self {
+            state: Arc::new(Mutex::new(SharedState {
+                selections: HashSet::new(),
+                values: Vec::new(),
+                on_change: None,
+            })),
+        }
+    }
+
+    // TODO: Handling of the global state
+
+    /// Adds a new checkbox to the group.
+    ///
+    /// The checkbox will display `label` next to it, and will ~embed~ `value`.
+    pub fn checkbox<S: Into<StyledString>>(&mut self, value: T, label: S) -> Checkbox
+    where
+        T: Send + Sync,
+    {
+        let element = Arc::new(value);
+        self.state.lock().values.push(element.clone());
+        Checkbox::labelled(label).on_change({
+            // TODO: consider consequences
+            let selectable = Arc::downgrade(&element);
+            let groupstate = self.state.clone();
+            move |_, checked| {
+                if checked {
+                    if let Some(v) = selectable.upgrade() {
+                        groupstate.lock().selections.insert(v);
+                    }
+                } else if let Some(v) = selectable.upgrade() {
+                    groupstate.lock().selections.remove(&v);
+                }
+            }
+        })
```

Comment:
If you don't need to keep the Arc around, it gets a bit simpler:
```suggestion
        let i = self.state.lock().add(value);
        
        Checkbox::labelled(label).on_change({
            let groupstate = Arc::clone(&self.state);
            move |_, checked| {
                groupstate.lock().set_checked(i, checked);
            }
        })
```